core/src/main/scala/quasar/datasource/kafka/FullConsumer.scala

```diff
@@ -0,0 +1,129 @@
+import cats.data.NonEmptySet
+import cats.effect._
+import cats.implicits._
+import fs2.Stream
+import fs2.kafka.{CommittableConsumerRecord, ConsumerSettings, KafkaConsumer, consumerResource}
+
+import scala.collection.immutable.SortedSet
+
+/**
+ * A [[Consumer]] that fetches all messages available in a topic up to the last message
+ * at the time it starts running.
+ */
+class FullConsumer[F[_]: ConcurrentEffect: ContextShift: Timer, K, V](
+    settings: ConsumerSettings[F, K, V],
+    decoder: RecordDecoder[F, K, V])
+    extends Consumer[F] with Logging {
+
+  import FullConsumer.topicPartitionOrder
+
+  val F: ConcurrentEffect[F] = ConcurrentEffect[F]
+
+  override def fetch(topic: String): Resource[F, Stream[F, Byte]] = {
+    consumerResource[F].using(settings) evalMap { consumer =>
+      for {
+        endOffsets <- assignNonEmptyPartitionsForTopic(consumer, topic)
+      } yield takeUntilEndOffsets(consumer.partitionedStream, endOffsets).flatMap(decoder)
+    }
+  }
+
+  /**
+   * Assigns all non-empty partitions of a topic to this consumer, and returns a map of
+   * end offsets for each one. The returned map will not contain offsets for empty partitions.
+   */
+  def assignNonEmptyPartitionsForTopic(
+      consumer: KafkaConsumer[F, K, V],
+      topic: String)
+      : F[Map[TopicPartition, Long]] = {
+    for {
+      info <- consumer.partitionsFor(topic)
+      topicPartitionSet = SortedSet(info.map(partitionInfoToTopicPartition): _*)
+      _ <- F.delay(log.info(s"TopicPartition Set: $topicPartitionSet"))
+      beginningOffsets <- consumer.beginningOffsets(topicPartitionSet)
+      _ <- F.delay(log.debug(s"${beginningOffsets.size} beginning offsets: $beginningOffsets"))
+      endOffsets <- consumer.endOffsets(topicPartitionSet)
+      _ <- F.delay(log.debug(s"${endOffsets.size} end offsets: $endOffsets"))
+      nonEmptyPartitions = topicPartitionSet.filterNot(tp => endOffsets(tp) == beginningOffsets(tp))
+      _ <- NonEmptySet.fromSet(nonEmptyPartitions).fold(F.unit)(consumer.assign)
+      _ <- F.delay(log.info(s"Assigned partitions ${nonEmptyPartitions.mkString(" ")}"))
+    } yield endOffsets.filterKeys(nonEmptyPartitions.contains)
+  }
+
+  def partitionInfoToTopicPartition(info: PartitionInfo): TopicPartition =
+    new TopicPartition(info.topic(), info.partition())
+
+  /**
+   * Emits all messages up to the last message of each partition given the map of end offsets.
+   *
+   * @param endOffsets Map of all _non-empty_ topic/partitions to their end offsets (offset of last message + 1)
+   */
+  def takeUntilEndOffsets(
+      // TODO: it would make more sense to take a KafkaConsumer, but that can't be mocked on fs2-kafka 1.0.0.
+      stream: Stream[F, Stream[F, CommittableConsumerRecord[F, K, V]]],
+      endOffsets: Map[TopicPartition, Long])
+      : Stream[F, CommittableConsumerRecord[F, K, V]] = {
+    stream
+      .take(endOffsets.size.toLong) // assumes no automatic assignment; needs a filter otherwise
```

Comment:
Is this comment valid anymore, now that we've filtered out the empty partitions elsewhere?

Reply:
Yes, it is valid. The assumption is valid here, but it won't be valid if we use `subscribe`, which we might do for offset-saving. Since this code is likely to be copy&pasted to implement that, I thought it would be helpful to point out the assumption.

Reply:
👍 makes sense!